azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/Timer.java

```diff
@@ -4,70 +4,60 @@
  */
 package com.microsoft.azure.eventhubs;
 
+import java.io.IOException;
 import java.time.Duration;
-import java.util.HashSet;
-import java.util.Locale;
-import java.util.concurrent.ScheduledFuture;
-import java.util.concurrent.ScheduledThreadPoolExecutor;
-import java.util.concurrent.TimeUnit;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.RejectedExecutionException;
 
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
+import com.microsoft.azure.eventhubs.amqp.DispatchHandler;
 
-/**
- * An abstraction for a Scheduler functionality - which can later be replaced by a light-weight Thread
- */
-public final class Timer {
-    private static ScheduledThreadPoolExecutor executor = null;
+final class Timer {
 
-    private static final Logger TRACE_LOGGER = LoggerFactory.getLogger(Timer.class);
-    private static final HashSet<String> references = new HashSet<String>();
-    private static final Object syncReferences = new Object();
+    final ISchedulerProvider schedulerProvider;
 
-    private Timer() {
+    public Timer(final ISchedulerProvider schedulerProvider) {
+        this.schedulerProvider = schedulerProvider;
     }
 
-    /**
-     * @param runFrequency implemented only for TimeUnit granularity - Seconds
-     */
-    public static ScheduledFuture<?> schedule(Runnable runnable, Duration runFrequency, TimerType timerType) {
-        switch (timerType) {
-            case OneTimeRun:
-                return executor.schedule(runnable, runFrequency.toMillis(), TimeUnit.MILLISECONDS);
-
-            case RepeatRun:
-                return executor.scheduleWithFixedDelay(runnable, runFrequency.toMillis(), runFrequency.toMillis(), TimeUnit.MILLISECONDS);
+    public CompletableFuture<?> schedule(
+            final Runnable runnable,
+            final Duration runAfter) {
 
-            default:
-                throw new UnsupportedOperationException("Unsupported timer pattern.");
+        final ScheduledTask scheduledTask = new ScheduledTask(runnable);
+        final CompletableFuture<?> taskHandle = scheduledTask.getScheduledFuture();
+        try {
+            this.schedulerProvider.getReactorScheduler().invoke((int) runAfter.toMillis(), scheduledTask);
+        } catch (IOException|RejectedExecutionException e) {
```

Comment:
To confirm, IOException and RejectedExecutionException are only exception which can be thrown before a task is executed, right?